File: conformance/chargebee/webhook.mjs
import {
  chargebee,
  createChargebeeWebhookProcessor,
} from "@chargebee/better-auth";
import Chargebee, { WebhookEventType } from "chargebee";
import { afterEach, describe, expect, test, vi } from "vitest";
import {
  callEndpoint,
  endpointContext,
  fakeAdapter,
  fakeChargebee,
  logger,
  plans,
  responseBody,
} from "./helpers.mjs";

const knownEvents = [
  "subscription_created",
  "subscription_activated",
  "subscription_changed",
  "subscription_renewed",
  "subscription_started",
  "subscription_cancelled",
  "subscription_scheduled_cancellation_removed",
  "customer_deleted",
];

function event(event_type, content = {}) {
  return { content, event_type, id: `event_${event_type}` };
}

function subscription(overrides = {}) {
  return {
    current_term_end: 1_800_000_000,
    current_term_start: 1_700_000_000,
    id: "provider_sub",
    status: "active",
    subscription_items: [{
      amount: 1200,
      item_price_id: "price_pro",
      item_type: "plan",
      quantity: 2,
      unit_price: 600,
    }],
    ...overrides,
  };
}

afterEach(() => {
  vi.restoreAllMocks();
});

describe("@chargebee/better-auth@1.2.0 webhook oracle", () => {
  test("registers exactly eight mappings, unhandled/error listeners, custom hook, and truthy-pair auth", async () => {
    for (const credentials of [
      {},
      { webhookUsername: "user" },
      { webhookPassword: "pass" },
      { webhookPassword: "pass", webhookUsername: "user" },
    ]) {
      const { client, handler } = fakeChargebee();
      const webhookHandler = vi.fn(instance => {
        expect(instance).toBe(handler);
        instance.on("custom_event", vi.fn());
      });
      const plugin = chargebee({ chargebeeClient: client, webhookHandler, ...credentials });
      const response = await callEndpoint(plugin.endpoints.chargebeeWebhook, {
        body: event("unhandled_oracle"),
      });
      expect(response.status).toBe(200);
      expect(await responseBody(response)).toEqual({ received: true });
      expect([...handler.listeners.keys()]).toEqual([
        ...knownEvents,
        "unhandled_event",
        "error",
        "custom_event",
      ]);
      const createOptions = client.webhooks.createHandler.mock.calls[0][0];
      if (credentials.webhookUsername && credentials.webhookPassword) {
        expect(createOptions.requestValidator).toBeTypeOf("function");
      } else {
        expect(createOptions.requestValidator).toBeUndefined();
      }
      expect(webhookHandler).toHaveBeenCalledBefore(handler.handle);
    }
    expect(WebhookEventType.SubscriptionScheduledCancellationRemoved)
      .toBe("subscription_scheduled_cancellation_removed");
  });

  test("documents the published invalid-auth acknowledgement bug that native code must harden", async () => {
    const client = new Chargebee({ apiKey: "oracle_key", site: "oracle-test" });
    const log = logger();
    const plugin = chargebee({
      chargebeeClient: client,
      webhookPassword: "correct-pass",
      webhookUsername: "correct-user",
    });
    const request = new Request("https://auth.example.test/api/auth/chargebee/webhook", {
      headers: {
        authorization: `Basic ${Buffer.from("wrong:credentials").toString("base64")}`,
      },
      method: "POST",
    });
    const response = await callEndpoint(plugin.endpoints.chargebeeWebhook, {
      body: event("subscription_created"),
      context: endpointContext({ logger: log }),
      request,
    });
    expect(response.status).toBe(200);
    expect(await responseBody(response)).toEqual({ received: true });
    expect(log.warn).toHaveBeenCalledWith(expect.stringContaining("Webhook rejected"));
  });

  test("documents that SDK handle resolves before async queue persistence", async () => {
    const client = new Chargebee({ apiKey: "oracle_key", site: "oracle-test" });
    let release;
    let persisted = false;
    const publish = vi.fn(() => new Promise(resolve => {
      release = () => {
        persisted = true;
        resolve();
      };
    }));
    const plugin = chargebee({
      chargebeeClient: client,
      webhookEventBus: { publish },
      webhookPassword: "pass",
      webhookUsername: "user",
    });
    const request = new Request("https://auth.example.test/api/auth/chargebee/webhook", {
      headers: {
        authorization: `Basic ${Buffer.from("user:pass").toString("base64")}`,
      },
      method: "POST",
    });
    const response = await callEndpoint(plugin.endpoints.chargebeeWebhook, {
      body: event("subscription_created", {
        customer: { id: "customer_1" },
        subscription: subscription(),
      }),
      request,
    });
    expect(publish).toHaveBeenCalledOnce();
    expect(persisted).toBe(false);
    expect(await responseBody(response)).toEqual({ received: true });
    release();
    await new Promise(setImmediate);
    expect(persisted).toBe(true);
  });

  test("created processing persists subscription/items sequentially then calls created and trial", async () => {
    const calls = [];
    const onSubscriptionCreated = vi.fn(async () => calls.push("created"));
    const onTrialStart = vi.fn(async () => calls.push("trial"));
    const adapter = fakeAdapter({
      create: vi.fn(async ({ data, model }) => {
        calls.push(`create:${model}`);
        return { id: model === "subscription" ? "local_created" : "item_created", ...data };
      }),
      findOne: vi.fn(async ({ model }) => model === "user" ? { id: "user_1" } : null),
    });
    const { client } = fakeChargebee();
    const processor = createChargebeeWebhookProcessor({
      chargebeeClient: client,
      subscription: { enabled: true, onSubscriptionCreated, onTrialStart, plans },
    }, { context: { adapter, logger: logger() } });
    await processor.process(event("subscription_created", {
      customer: { id: "customer_1" },
      subscription: subscription({
        meta_data: { subscriptionId: "metadata_pending" },
        trial_end: 1_700_100_000,
        trial_start: 1_700_000_000,
      }),
    }));
    expect(calls).toEqual(["create:subscription", "create:subscriptionItem", "created", "trial"]);
  });

  test("completion lookup updates without replacing items and orders trial before completion", async () => {
    const calls = [];
    const local = { id: "local_pending", referenceId: "user_1", status: "future" };
    const adapter = fakeAdapter({
      create: vi.fn(async () => { calls.push("create"); }),
      deleteMany: vi.fn(async () => { calls.push("delete"); }),
      findOne: vi.fn()
        .mockResolvedValueOnce(null)
        .mockResolvedValueOnce(null)
        .mockResolvedValueOnce(local),
      update: vi.fn(async ({ update }) => ({ ...local, ...update })),
    });
    const { client } = fakeChargebee();
    const processor = createChargebeeWebhookProcessor({
      chargebeeClient: client,
      subscription: {
        enabled: true,
        onSubscriptionComplete: vi.fn(async () => calls.push("complete")),
        onTrialStart: vi.fn(async () => calls.push("trial")),
        plans,
      },
    }, { context: { adapter, logger: logger() } });
    await processor.process(event("subscription_activated", {
      customer: { id: "customer_1", meta_data: { pendingSubscriptionId: "local_pending" } },
      subscription: subscription({
        meta_data: { subscriptionId: "metadata_pending" },
        trial_end: 1_700_100_000,
        trial_start: 1_700_000_000,
      }),
    }));
    expect(adapter.findOne.mock.calls.map(([value]) => value.where[0])).toEqual([
      { field: "chargebeeSubscriptionId", value: "provider_sub" },
      { field: "id", value: "metadata_pending" },
      { field: "id", value: "local_pending" },
    ]);
    expect(calls).toEqual(["trial", "complete"]);
    expect(adapter.deleteMany).not.toHaveBeenCalled();
    expect(adapter.create).not.toHaveBeenCalled();
  });

  test("update replaces items then orders newly-cancelled, update, and trial-end callbacks", async () => {
    const calls = [];
    const local = {
      canceledAt: null,
      id: "local_active",
      referenceId: "user_1",
      status: "in_trial",
    };
    const adapter = fakeAdapter({
      create: vi.fn(async ({ model }) => calls.push(`create:${model}`)),
      deleteMany: vi.fn(async () => calls.push("delete-items")),
      findOne: vi.fn(async () => local),
      update: vi.fn(async ({ update }) => ({ ...local, ...update })),
    });
    const { client } = fakeChargebee();
    const processor = createChargebeeWebhookProcessor({
      chargebeeClient: client,
      subscription: {
        enabled: true,
        onSubscriptionCancel: vi.fn(async () => calls.push("cancel")),
        onSubscriptionUpdate: vi.fn(async () => calls.push("update")),
        onTrialEnd: vi.fn(async () => calls.push("trial-end")),
        plans,
      },
    }, { context: { adapter, logger: logger() } });
    await processor.process(event("subscription_scheduled_cancellation_removed", {
      customer: { id: "customer_1" },
      subscription: subscription({ cancelled_at: 1_700_050_000, status: "active" }),
    }));
    expect(calls).toEqual([
      "delete-items",
      "create:subscriptionItem",
      "cancel",
      "update",
      "trial-end",
    ]);
  });

  test("customer deletion removes rows/items and clears metadata plus configured fallback", async () => {
    const adapter = fakeAdapter({
      findMany: vi.fn(async ({ model }) => {
        if (model === "subscription") return [{ id: "local_1" }, { id: "local_2" }];
        if (model === "user") return [{ id: "fallback_user" }];
        return [];
      }),
    });
    const log = logger();
    const { client } = fakeChargebee();
    const processor = createChargebeeWebhookProcessor({ chargebeeClient: client }, {
      auth: { $context: Promise.resolve({ adapter, logger: log }) },
    });
    await processor.process(event("customer_deleted", {
      customer: { id: "customer_1", meta_data: { customerType: "user", userId: "metadata_user" } },
    }));
    expect(adapter.deleteMany).toHaveBeenCalledTimes(4);
    expect(adapter.update.mock.calls).toEqual([
      [{ model: "user", update: { chargebeeCustomerId: null }, where: [{ field: "id", value: "metadata_user" }] }],
      [{ model: "user", update: { chargebeeCustomerId: null }, where: [{ field: "id", value: "fallback_user" }] }],
    ]);
  });
});

File: tests/chargebee_contract/routes/webhook.rs
use super::super::support::{fixture, post, raw_post};
use async_trait::async_trait;
use axum::http::StatusCode;
use lucid_auth::{
    ChargebeeCallbackError, ChargebeeStore, ChargebeeSubscriptionStatus, ChargebeeWebhookEvent,
    ChargebeeWebhookEventBus,
};
use serde_json::json;
use std::sync::Arc;

struct FailingBus;

#[async_trait]
impl ChargebeeWebhookEventBus for FailingBus {
    async fn publish(&self, _event: ChargebeeWebhookEvent) -> Result<(), ChargebeeCallbackError> {
        Err(ChargebeeCallbackError::new("queue implementation detail"))
    }
}

#[tokio::test]
async fn configured_basic_auth_rejects_invalid_credentials_but_payload_errors_acknowledge() {
    let secured = fixture(false, |options| {
        options.webhook_username = Some("user".into());
        options.webhook_password = Some("pass".into());
    })
    .await;
    let event = json!({"event_type": "unhandled_event", "id": "evt", "content": {}}).to_string();
    let (status, _) = raw_post(
        &secured,
        "/api/auth/chargebee/webhook",
        &event,
        Some("Basic invalid"),
    )
    .await;
    assert_eq!(status, StatusCode::UNAUTHORIZED);

    let (status, body) = raw_post(
        &secured,
        "/api/auth/chargebee/webhook",
        "not-json",
        Some("Basic dXNlcjpwYXNz"),
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    assert_eq!(body, json!({"received": true}));
}

#[tokio::test]
async fn webhook_bus_is_awaited_and_failure_uses_the_public_message() {
    let fixture = fixture(false, |options| {
        options.webhook_event_bus = Some(Arc::new(FailingBus));
    })
    .await;
    let event =
        json!({"event_type": "subscription_created", "id": "evt", "content": {}}).to_string();
    let (status, body) = raw_post(&fixture, "/api/auth/chargebee/webhook", &event, None).await;
    assert_eq!(status, StatusCode::INTERNAL_SERVER_ERROR);
    assert_eq!(body["message"], "Failed to queue webhook event");
}

#[tokio::test]
async fn activated_webhook_promotes_pending_without_creating_item_rows() {
    let fixture = fixture(true, |_| {}).await;
    let (status, _) = post(
        &fixture,
        "/api/auth/subscription/create",
        json!({
            "itemPriceId": "price_pro",
            "successUrl": "/success",
            "cancelUrl": "/cancel",
            "disableRedirect": true
        }),
    )
    .await;
    assert_eq!(status, StatusCode::OK);
    let reference = fixture.user_id.unwrap().to_string();
    let pending = fixture
        .store
        .list_subscriptions_by_reference(&reference)
        .await
        .unwrap()
        .remove(0);
    let event = json!({
        "event_type": "subscription_activated",
        "id": "event_activated",
        "content": {
            "customer": {
                "id": "customer_contract",
                "meta_data": {"pendingSubscriptionId": pending.id.to_string()}
            },
            "subscription": {
                "id": "subscription_provider",
                "customer_id": "customer_contract",
                "status": "active",
                "current_term_start": 1700000000,
                "current_term_end": 1702592000,
                "subscription_items": [{
                    "item_price_id": "price_pro",
                    "item_type": "plan",
                    "quantity": 4,
                    "unit_price": 2500,
                    "amount": 10000
                }]
            }
        }
    })
    .to_string();
    let (status, body) = raw_post(&fixture, "/api/auth/chargebee/webhook", &event, None).await;
    assert_eq!(status, StatusCode::OK, "{body}");
    let updated = fixture
        .store
        .find_subscription(pending.id)
        .await
        .unwrap()
        .unwrap();
    assert_eq!(updated.status, ChargebeeSubscriptionStatus::Active);
    assert_eq!(
        updated.chargebee_subscription_id.as_deref(),
        Some("subscription_provider")
    );
    assert_eq!(updated.seats, Some(4.0));
    let items = fixture
        .store
        .list_subscription_items(pending.id)
        .await
        .unwrap();
    assert!(items.is_empty());
}
